include/nbl/asset/filters/CNormalMapToDerivativeFilter.h

```diff
@@ -0,0 +1,275 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef __NBL_ASSET_C_NORMAL_MAP_TO_DERIVATIVE_FILTER_H_INCLUDED__
+#define __NBL_ASSET_C_NORMAL_MAP_TO_DERIVATIVE_FILTER_H_INCLUDED__
+
+#include "nbl/core/core.h"
+
+#include <type_traits>
+#include <functional>
+ 
+#include "nbl/asset/filters/CMatchedSizeInOutImageFilterCommon.h"
+#include "CConvertFormatImageFilter.h"
+
+namespace nbl
+{
+namespace asset
+{
+
+class CNormalMapToDerivativeFilterBase
+{
+	public:
+		class CNormalMapToDerivativeStateBase
+		{
+			public:
+
+				static inline constexpr size_t decodeTypeByteSize = sizeof(double);
+				static inline constexpr size_t forcedScratchChannelAmount = 4;
+				uint8_t*	scratchMemory = nullptr;										//!< memory covering all regions used for temporary filling within computation of sum values
+				size_t	scratchMemoryByteSize = {};											//!< required byte size for entire scratch memory
+
+				static inline size_t getRequiredScratchByteSize(asset::VkExtent3D extent)
+				{
+					size_t retval = extent.width * extent.height * extent.depth * decodeTypeByteSize * forcedScratchChannelAmount;
+					
+					return retval;
+				}
+
+				/*
+					Layer ID is relative to outBaseLayer in state
+				*/
+
+				const std::array<double, forcedScratchChannelAmount>& getAbsoluteLayerScaleValue(size_t layer)
+				{
+					if (!maxAbsLayerScaleValues.empty())
+						return maxAbsLayerScaleValues[layer];
+					else
+						return {};
+				}
+
+			protected:
+				std::vector<std::array<double, forcedScratchChannelAmount>> maxAbsLayerScaleValues; 								//!< scales gained by the filter (each layer handled) for derivative map shader usage
+		};
+
+	protected:
+		CNormalMapToDerivativeFilterBase() {}
+		virtual ~CNormalMapToDerivativeFilterBase() {}
+
+		static inline bool validate(CNormalMapToDerivativeStateBase* state)
+		{
+			if (!state)
+				return false;
+
```

Comment:
probably validate that scratch exists and the memory reserved is not 0-sized

Reply:
corrected in following commit